couchbase/cluster.hxx

```diff
@@ -99,29 +97,34 @@ class cluster
     template<typename Handler>
     void open_bucket(const std::string& bucket_name, Handler&& handler)
     {
-        if (buckets_.find(bucket_name) != buckets_.end()) {
+        auto b = find_bucket_by_name(bucket_name);
+        if (b) {
             return handler({});
         }
+
         std::vector<protocol::hello_feature> known_features;
         if (session_ && session_->has_config()) {
             known_features = session_->supported_features();
         }
-        auto b = std::make_shared<bucket>(id_, ctx_, tls_, tracer_, meter_, bucket_name, origin_, known_features);
+        b = std::make_shared<bucket>(id_, ctx_, tls_, tracer_, meter_, bucket_name, origin_, known_features);
         b->bootstrap([this, handler = std::forward<Handler>(handler)](std::error_code ec, const topology::configuration& config) mutable {
             if (!ec && !session_->supports_gcccp()) {
                 session_manager_->set_configuration(config, origin_.options());
             }
             handler(ec);
         });
+        std::scoped_lock lock(buckets_mutex_);
         buckets_.emplace(bucket_name, b);
     }
 
     template<typename Handler>
     void close_bucket(const std::string& bucket_name, Handler&& handler)
     {
-        auto ptr = buckets_.find(bucket_name);
-        if (ptr != buckets_.end()) {
-            ptr->second->close();
+        auto bucket = find_bucket_by_name(bucket_name);
```

Comment:
Similarly to the comment above, we need to acquire the lock, find the bucket, then remove it from the list and then unlock the lock, then finally close the bucket instance.  Otherwise there is a chance that the bucket gets erased and closed twice.